packages/client/src/bags/types/bagOperation.types.ts

```diff
@@ -0,0 +1,177 @@
+  oldValue?: OperationChange['oldValue'];
+  newValue?: OperationChange['newValue'];
+};
+
+type ChangeTypeItemOriginChanged = {
+  changeType: ChangeType.ItemOriginChanged;
+  productId?: OperationChange['productId'];
+  variantId?: OperationChange['variantId'];
+};
+
+type ChangeTypeItemPriceChanged = {
+  changeType: ChangeType.ItemPriceChanged;
+  itemId?: OperationChange['itemId'];
+  newValue?: OperationChange['newValue'];
+  oldValue?: OperationChange['oldValue'];
+  productId?: OperationChange['productId'];
+  variantId?: OperationChange['variantId'];
+};
+
+type ChangeTypeItemQuantityChanged = {
+  changeType: ChangeType.ItemQuantityChanged;
+  itemId?: OperationChange['itemId'];
+  newValue?: OperationChange['newValue'];
+  oldValue?: OperationChange['oldValue'];
+};
+
+type ChangeTypeItemNotEnoughStock = {
+  changeType: ChangeType.NotEnoughStock;
+  itemId?: OperationChange['itemId'];
+  stockAvailable?: OperationChange['stockAvailable'];
+};
+
+type ChangeTypePromoCodeEligibilityChanged = {
+  changeType: ChangeType.PromoCodeEligibilityChanged;
+};
+
+type ChangeTypePromotionNotApplied = {
+  changeType: ChangeType.PromotionNotApplied;
+};
+
+type ChangeTypeShippingCostChanged = {
+  changeType: ChangeType.ShippingCostChanged;
+};
+
+type ChangeTypeShippingFlatRateEligibilityChanged = {
+  changeType: ChangeType.ShippingFlatRateEligibilityChanged;
+};
+
+type ChangeTypeTotalDiscountChanged = {
+  changeType: ChangeType.TotalDiscountChanged;
+  newValue?: OperationChange['newValue'];
+  oldValue?: OperationChange['oldValue'];
+};
+
+type ChangeTypeSelectedSaleIntentChanged = {
+  changeType: ChangeType.SelectedSaleIntentChanged;
+  productId?: OperationChange['productId'];
+  variantId?: OperationChange['variantId'];
+  itemId?: OperationChange['itemId'];
+  oldValue?: OperationChange['oldValue'];
+  newValue?: OperationChange['newValue'];
+};
+
+export type BagOperationChange =
+  | ChangeTypeItemDiscountChanged
+  | ChangeTypeItemIsNowAvailable
+  | ChangeTypeItemIsUnavailable
+  | ChangeTypeItemMerged
+  | ChangeTypeItemOriginChanged
+  | ChangeTypeItemPriceChanged
+  | ChangeTypeItemQuantityChanged
+  | ChangeTypeItemNotEnoughStock
+  | ChangeTypePromoCodeEligibilityChanged
+  | ChangeTypePromotionNotApplied
+  | ChangeTypeSelectedSaleIntentChanged
+  | ChangeTypeShippingCostChanged
+  | ChangeTypeShippingFlatRateEligibilityChanged
+  | ChangeTypeTotalDiscountChanged;
+
+export enum BagViolationFixSuggestionAction {
+  DecreaseQuantity = 'DecreaseQuantity',
+  IncreaseValue = 'IncreaseValue',
+  DecreaseValue = 'DecreaseValue',
+}
+
+type BagViolationFixSuggestion = {
+  action: BagViolationFixSuggestionAction;
+  value: string;
+};
+
+export type BagViolation = {
+  itemsIds?: Array<BagItem['id']>;
+  fixSuggestion?: BagViolationFixSuggestion;
+};
+
+export type BagOperation = {
+  id: string;
+  createdDate: string;
+  changes: Array<BagOperationChange>;
+  violations: Array<BagViolation>;
```

Comment:
Shouldn't we use the `CheckoutViolation` here? (Although I also have strange feelings about this, this comment is more of an attempt to mimic the API)

Reply:
It didn't feel right. I created a new type with the same properties.